app/src/androidTest/java/sdp/moneyrun/DatabaseProxyTest.java

```diff
@@ -42,4 +53,121 @@ public void onComplete(@NonNull Task<DataSnapshot> task) {
         }
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void addPlayerListenerThrowsExceptionOnNullPlayer() {
+        DatabaseProxy db = new DatabaseProxy();
+        ValueEventListener listener = new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+                return;
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {
+
+            }
+        };
+        db.addPlayerListener(null, listener);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void addPlayerListenerThrowsExceptionOnNullListener() {
+        DatabaseProxy db = new DatabaseProxy();
+        Player player = new Player(1);
+        db.addPlayerListener(player, null);
+    }
+
+    @Test
+    public void addPlayerListenerCorrectlyUpdatesData(){
+        Player player = new Player(564123);
+        player.setAddress("FooBarr");
+        player.setName("Johann");
+        final DatabaseProxy db = new DatabaseProxy();
+        db.putPlayer(player);
+        String newName = "Simon";
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        db.addPlayerListener(player, new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+                Player p = snapshot.getValue(Player.class);
+                player.setName(p.getName());
+
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {
+                assert(false);
+            }
+        });
+        Player p = new Player(564123,newName,"FooBarr",0,0);
```

Comment:
It depends on the Listener you add to the function. Basically the onDataChange function of the listener will be triggered every time the data about that player in the database is changed. What is done at that point is up to the caller

Reply:
If I add this line of code right after line 106 : `Player p2 = new Player(564123,"Rodrigo","FooBarr",0,0);` will the name of the Player object player change?

Reply:
No because you haven't put it in the database
But if you put it on the database, the onDataChange method will be triggered. In the case of my test, the only thing it does is change the name

Reply:
So if we put p2 in the database, name field of player will change?

Reply:
yes

Reply:
But then if we have 2 Player objects that we want to attach a listener to ( say player1 and player 2), would this be a problem? I mean would this mean that putting p1 in the database updates both player1 and player2 name fields? Even if they had unique names at first? Like imagine having lot of players and we want to track their evolution so we put listeners on all of them, having one player put into the database with new name(like p1 in your test) we would want to update only one of the players not all of them is this guaranteed or not?

Reply:
Sorry for the misunderstanding, the listener is added to a db entry not an instance of player. Which instance(s) of player you want to update when that entry is modified is up to the caller which will do that in the listener he must provide as an argument

Reply:
Ohhhh, I think I got it if we have listeners on all players but put p1 in the database it will be updated with respect to playerId, yeah so obvious didn't realize they were the same yes I get it now